### ISSUE TITLE
[Bug] Crafter is being seen as a valid destination when it should not

### ISSUE THREAD

**Issue by Krougal** [Astavie/ThermalLogistics#43]
So dump to network from Terminal, or a servo on any inventory sees a crafter as a valid destination if the materials match. This is much more serious than it might seem. I am guessing I didn't notice this before since they weren't shortest destination.

Crafter should only be taking materials when a job is requested from them, and only the amounts needed. It is really bad on devices with multiple recipes (like smelter/pulverizer)

Actually, it looks like it is being seen as a valid destination for anything dumped into the network from the Terminal in particular, since things will go jam there. For instance I don't have any storage for sticks, if I dump sticks back they will go out and jam a crafter. Before I had the crafters in the sticks would not leave the terminal.

**Comment by Krougal** [comment]
Even more confusing is the system started behaving again (although I did put in dense itemducts) as far as the terminal is not pushing junk.

It seems however that upon creating a new recipe in a crafter, the network decides it should go and push a batch of it to celebrate. Generally from 8-32 items (seems like enough to make 8 batches). the only servo I have left in open network is for fluxed phyto-gro, which thankfully is not being pushed. Crafters are resonant, which would pull a stack by default I would think (unless you made that change we talked about).

**Comment by Krougal** [comment]
So definite 2 part bug. The consistent part is when I lay out new recipes, it wants to make 8 batches of each new recipe. I've only noticed this in my survival game. Persisting after reload. The other part is the intermittent dumping of invalid junk into the network targeted at the crafters (from Terminal dump or servo/distributor)

**Comment by Krougal** [comment]
The severe retardation continues all day long. This is in my survival game, I do not know if it is because of other mods or I just did not notice in creative. I have several recipes that are intermittent, and I don't know if it is a routing thing (it should not be because everything is being done by crafters pulling) or other mods or what, but I'm about ready to throw in the towel and go to RS.

For example. If I try to make a signalum upgrade from the terminal and request the electrum gear (I have all necessary smelters and crafters in place, plenty of stock) sometimes 2 electrum will get made, sometimes none, sometimes they go to the gearpress, sometimes they wind up in the pulverizer and then down in the transposer like they were gonna make fluxed electrum, except no destablized redstone gets requested from the crucible either.

**Comment by Krougal** [comment]
I can only assume this is a feature since it is regular, place crafter, add recipes, wait a little while, crafter will try to make 8 of each recipe. This causes a lot of undesired behavior.

I'm also having trouble with sourcing. For example, I have my "misc crap" strongbox for items I don't have a place for really. It has a filter to keep from getting filled. I tend to keep excess alloys, etc there. Sometimes a crafting process (or request from terminal) will see and pull material from there just fine, other times it goes ahead and has them crafted by another crafter instead of using these. It doesn't seem to matter how many are present or if they are listed items on filter or not.

Then we've got the issue with multiple output items; I need 3 signalum but it is produced in batches of 4. If I am requesting it myself I make sure to request items in the right multiples to work around this, but when it's triggered by a requester I can't control that.

The destination/jamming thing seems to be caused by the crafter being "open" so it is waiting for whatever it asked for and for some reason the servo pushing phyto-gro (yeah I really need to put requesters or make it a private line) decides it should send some. The dump chest also gets confused by this, since items that don't have a place (and part of the use of the dump chest is finding what you still need to add storage for) will sit there until a crafter fires off and then try to go there. 

Lastly, what is the routing for requests? I assumed that it would prefer existing material over triggering another crafter. I am starting to think it goes by nearest or round-robin because the results are inconsistent. Multiple sources also seem to not work (needs 4 of something, there are already 2 so it should use those and then request 2 more be crafted).

**Comment by Astavie** [comment]
I have a theory: since it's so regular, do you have a basic requester somewhere set to blacklist? Basic requesters request 8 at a time and requesters set to blacklist will literally request from everything

**Comment by Krougal** [comment]
> I have a theory: since it's so regular, do you have a basic requester somewhere set to blacklist? Basic requesters request 8 at a time and requesters set to blacklist will literally request from everything

So you're saying this isn't by design.
I would sure hope not, and I'd think I would know for sure since it would cause something else to break. I will double check though.
I can't imagine a case where I would set a requester to blacklist instead of whitelist.

**Comment by Astavie** [comment]
Also, requesters will try to pull from existing items first and crafters second. However, if for whatever reason it fails to pull, it immediately tries requesting from crafters. This is kinda a bug but also exactly as intended (though it indeed has unintended consequences). I'm trying to fix this in the upcoming 0.3 update (where I'm rewriting the broken request algorithm completely) but that won't be for at least a few weeks

**Comment by Krougal** [comment]
Cool. In the meantime I'll continue refining what I can from my end. Just like all engineering, what worked fine in the lab (creative) doesn't in the real world (survival) so we have to go re-engineer.

There are some things that are not the fault of TL, for instance the alloy induction smelter has the same issue as the fluid transposer bucket recipes, because I have to unlock the flux slot, it makes some of the outputs valid inputs. In this case the buffer prevents some issues (over-sending) but enables others (this one). Sometimes the machine input fires before the crafter can send it on it's way (usually when it is the last craft requested). This is because of the hardened metal glass recipes. So I realized I need to put a filtered duct from the buffer to the smelter to prevent this. The other issue is being careful about starting multiple requests at the same time, say electrum and invar, because if the iron and silver get in at the same time we have a jam (and this is buffer or no buffer). 

Actually that does bring up another suggestion (sorry this is turning into a crafter laundry list), some kind of check that previous job was finished before we start the next? I'm not sure how viable that is or that it won't cause other issues (could interfere with linked crafters?), needs to be thought through. Possibly just a small delay before we start the next input pull.

As I've learned, servos pretty much don't belong in this environment, re-routing the phyto-gro will at least reduce the amount of jams until your rewrite.

**Comment by Krougal** [comment]
Still working on re-organizing to get rid of that servo (I need to move a lot of stuff) but I cut it to 2 output (from 8) so there is less chance of it over-sending (hopefully it keeps up with demand). That and being very careful about what I dump back to network (and always using a dumpchest with distributor instead of the terminal) has at least eliminated most jams.

The changes to the smelter buffer have stopped internal screw-ups (although the item allocator was a disaster...it can't accept say silver and gold at the same time, I had to go back to strongbox) but requesting is still iffy. It seems like orders don't get completed, so if I ask for say 8 electrum, I get 6. The last 2 just sits there in queue. Same for other orders (mostly requested by other crafters), so if I ask for 8 Signalum ingots the pulverizer might not get all the copper or skip the silver, the smelter seems to forget to request sand for the 2nd batch. Sometimes I see ingots go from storage and get sent back, sometimes not. Maybe it is a matter of multiple inputs hitting the crafter at the same time? I've checked and rechecked that everything is setup properly. It all works fine until it doesn't. 

I think adding a small delay (to account for ejection cycles not being instant, especially on devices and low tier machines/servos/distributors) between requesting from stock and requesting from crafter would go a long way towards resolving this (and maybe that is what you had in mind already for your re-write)

Hopefully your rewrite takes care of these issues, I think the system will be pretty rock solid at that point.
